### DISCUSSION TITLE
Why do I get inconsistent query execution time with huge delays?

### DISCUSSION THREAD

**Discussion by Deliaz** [prisma/prisma#22070] [Q&A] [answered]
### Question

I am prototyping a web application and noticed that it takes a lot of time to create 10-15 records in just a few tables. 
Added custom metrics, basically measuring via `performance.now()`; 

Here are some examples: 
<kbd>
<img width="400" alt="Screenshot 2023-11-22 at 13 15 05" src="https://github.com/prisma/prisma/assets/4865430/e40804c5-70c3-409e-9ea3-001978dae4d3">
</kbd>
<kbd>
<img width="400" style="border: 2px solid  gray;" alt="Screenshot 2023-11-22 at 17 51 12" src="https://github.com/prisma/prisma/assets/4865430/28e42217-5a33-4eda-b438-bca870297f97">
</kbd>

It is not always the same procedure. But it can be for any element, any writing procedure. 

Tried with
* supabase
* vercel db
* digital ocean docker container
Numbers vary a bit, but the problem happens everywhere except postgres container @ localhost. 

How can I investigate this problem? 

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
model DomainActivity {
  id                       String            @id @default(uuid())
  date                     DateTime          @db.Timestamptz()
  domainId                 String
  memberToken              String
  member                   Member            @relation(fields: [memberToken], references: [token], onDelete: Cascade)
  timeSpent                Int
  sessionActivities        SessionActivity[]
  activitiesCount          Int
  lastSessionEndDatetime   DateTime?
  lastSessionStartDatetime DateTime?

  @@unique([date, domainId, memberToken])
  @@index([date, domainId, memberToken])
}
...
model Domain {
  id         String     @id @default(uuid())
  domain     String     @unique
  categories Category[]
}
```

```ts
...
  // Update a domain activity record.
  await prismadb.domainActivity.update({
    where: {
      date_domainId_memberToken: {
        domainId: domainRecord.id,
        memberToken: token,
        date,
      },
    },
    data: {
      timeSpent: {
        increment: timeSpentInc,
      },
      activitiesCount: {
        increment: sessionRecords.count,
      },
      lastSessionEndDatetime: new Date(lastSessionEndTimeTs),
      lastSessionStartDatetime: new Date(lastSessionStartTimeTs),
    },
  });

...

async function upsertDomain(domain: string) {
  // Check if the domain already exists
  const existingDomain = await prismadb.domain.findUnique({
    where: {
      domain,
    },
  });

  // If the domain does not exist, create it
  if (!existingDomain) {
    return prismadb.domain.create({
      data: {
        domain,
      },
    });
  }

  return existingDomain;
}

```

- OS: macOS, Debian, Ubuntu
- Database: PostgreSQL
- Node.js version: v18
- Prisma: 5.6.0

**Comment by ludralph** [comment]
Hi @Deliaz 👋 

Thank you for raising this question.

Network latency could be a factor when connecting to databases hosted on services like Supabase, Vercel DB, or Digital Ocean. The physical distance between your application server and the database server can affect response times. Is your application hosted in the same region as your DB? 

Another thing to note is that the first query might take longer due to the time required to establish a TCP connection to the database.

**Reply by Deliaz** [reply, marked as answer]
Hi @ludralph, thanks for the reply. 
I'm almost sure that even a bad latency can't cause minutes (sometimes up to 5 minutes) of delay. 
As you can see in the screenshot, it usually happens with a single but random operation in a loop. 

I tried different regions and the problem was almost the same. 

It turned out that the problem was with Vercel itself. 
I moved the project to a self-configured VPS server and now every query batch takes only 300-500ms with remote database. The problem is gone. 

I do not know the exact reason, but it could be related to execution limits in a free tier, or somehow related to a serverless function cold start. 
I have reported the issue and I hope they will take a look at it. 

Thanks again, 
I hope this thread is helpful for people struggling with the same issue.
